File: tests/array.rs
use rjvm::vm::data_type::Value::{Double, Float, Int, Long};

#[path = "./java_utils.rs"]
mod java;

#[test]
fn short_array() {
    assert_eq!(java::run_method("test_data/Array", "short_array"), Int(3));
}

#[test]
fn int_array() {
    assert_eq!(java::run_method("test_data/Array", "int_array"), Int(3));
}

#[test]
fn long_array() {
    assert_eq!(java::run_method("test_data/Array", "long_array"), Long(3));
}

#[test]
fn byte_array() {
    assert_eq!(java::run_method("test_data/Array", "byte_array"), Int(0xff));
}

#[test]
fn char_array() {
    assert_eq!(
        java::run_method("test_data/Array", "char_array"),
        Int('b' as i32)
    );
}

#[test]
fn float_array() {
    assert_eq!(
        java::run_method("test_data/Array", "float_array"),
        Float(-8.0)
    );
}

#[test]
fn double_array() {
    assert_eq!(
        java::run_method("test_data/Array", "double_array"),
        Double(8.0)
    );
}

#[test]
fn array_length() {
    assert_eq!(
        java::run_method("test_data/Array", "array_length"),
        Int(100)
    );
}

/* TODO implement after exceptions
#[test]
fn array_length_npe() {
    assert_eq!(java::run_method("test_data/Array", "array_length_npe"), Int(100));
}*/

File: tests/test_data/Array.java
package test_data;


class Array {

    public static int int_array() {
        int[] i = new int[10];
        i[0] = 1;
        i[1] = i[0] + 1;

        return i[0] + i[1];
    }

    public static byte byte_array() {
        byte[] b = new byte[10];
        b[1] = 0x0f;
        b[2] = (byte)0xf0;
        b[3] = (byte)(b[1] | b[2]);

        return b[3];
    }

    public static char char_array() {
        char[] b = new char[10];
        b[1] = 'a';
        b[2] = (char)(b[1] + 1);

        return b[2];
    }

    public static int array_length() {
        int[] f = new int[100];
        return f.length;
    }

    public static int array_length_npe() {
        int[] f = null;
        return f.length;
    }

    // TODO different types of arrays
}
